File: server/internal/protocol/orchestration_models.go
package protocol

import "encoding/json"

// OrchestrationReadModel is the top-level snapshot returned by getSnapshot.
type OrchestrationReadModel struct {
	SnapshotSequence int                    `json:"snapshotSequence"`
	Projects         []OrchestrationProject `json:"projects"`
	Threads          []OrchestrationThread  `json:"threads"`
	UpdatedAt        string                 `json:"updatedAt"`
}

// OrchestrationProject represents a workspace project.
type OrchestrationProject struct {
	ID                    ProjectId       `json:"id"`
	Title                 string          `json:"title"`
	WorkspaceRoot         string          `json:"workspaceRoot"`
	DefaultModelSelection *ModelSelection `json:"defaultModelSelection,omitempty"`
	Scripts               []ProjectScript `json:"scripts"`
	CreatedAt             string          `json:"createdAt"`
	UpdatedAt             string          `json:"updatedAt"`
	DeletedAt             string          `json:"deletedAt,omitempty"`
}

// ProjectScript is a user-defined script attached to a project.
type ProjectScript struct {
	ID                  string `json:"id"`
	Name                string `json:"name"`
	Command             string `json:"command"`
	Icon                string `json:"icon"`
	RunOnWorktreeCreate bool   `json:"runOnWorktreeCreate"`
}

// OrchestrationThread represents a conversation thread.
type OrchestrationThread struct {
	ID              ThreadId                         `json:"id"`
	ProjectId       ProjectId                        `json:"projectId"`
	Title           string                           `json:"title"`
	ModelSelection  ModelSelection                   `json:"modelSelection"`
	RuntimeMode     string                           `json:"runtimeMode"`
	InteractionMode string                           `json:"interactionMode"`
	Branch          *string                          `json:"branch,omitempty"`
	WorktreePath    *string                          `json:"worktreePath,omitempty"`
	LatestTurn      *OrchestrationLatestTurn         `json:"latestTurn,omitempty"`
	Messages        []OrchestrationMessage           `json:"messages"`
	ProposedPlans   []OrchestrationProposedPlan      `json:"proposedPlans"`
	Activities      []OrchestrationThreadActivity    `json:"activities"`
	Checkpoints     []OrchestrationCheckpointSummary `json:"checkpoints"`
	Session         *OrchestrationSession            `json:"session,omitempty"`
	CreatedAt       string                           `json:"createdAt"`
	UpdatedAt       string                           `json:"updatedAt"`
	DeletedAt       string                           `json:"deletedAt,omitempty"`
}

// OrchestrationMessage is a single message in a thread.
type OrchestrationMessage struct {
	ID          MessageId        `json:"id"`
	Role        string           `json:"role"`
	Text        string           `json:"text"`
	Attachments []ChatAttachment `json:"attachments"`
	TurnId      *TurnId          `json:"turnId,omitempty"`
	Streaming   bool             `json:"streaming"`
	CreatedAt   string           `json:"createdAt"`
	UpdatedAt   string           `json:"updatedAt"`
}

// ChatAttachment is a file attachment on a message.
type ChatAttachment struct {
	Type      string `json:"type"`
	ID        string `json:"id"`
	Name      string `json:"name"`
	MimeType  string `json:"mimeType"`
	SizeBytes int    `json:"sizeBytes"`
}

// OrchestrationLatestTurn tracks the most recent turn in a thread.
type OrchestrationLatestTurn struct {
	TurnId             TurnId     `json:"turnId"`
	State              string     `json:"state"`
	RequestedAt        string     `json:"requestedAt"`
	StartedAt          *string    `json:"startedAt,omitempty"`
	CompletedAt        *string    `json:"completedAt,omitempty"`
	AssistantMessageId *MessageId `json:"assistantMessageId,omitempty"`
}

// OrchestrationSession describes the runtime session for a thread.
type OrchestrationSession struct {
	ThreadId     ThreadId `json:"threadId"`
	Status       string   `json:"status"`
	ProviderName *string  `json:"providerName,omitempty"`
	RuntimeMode  string   `json:"runtimeMode"`
	ActiveTurnId *TurnId  `json:"activeTurnId,omitempty"`
	LastError    *string  `json:"lastError,omitempty"`
	UpdatedAt    string   `json:"updatedAt"`
}

// OrchestrationProposedPlan is a plan the assistant has proposed.
type OrchestrationProposedPlan struct {
	ID                     string    `json:"id"`
	TurnId                 *TurnId   `json:"turnId,omitempty"`
	PlanMarkdown           string    `json:"planMarkdown"`
	ImplementedAt          *string   `json:"implementedAt,omitempty"`
	ImplementationThreadId *ThreadId `json:"implementationThreadId,omitempty"`
	CreatedAt              string    `json:"createdAt"`
	UpdatedAt              string    `json:"updatedAt"`
}

// OrchestrationCheckpointSummary describes a checkpoint in a thread.
type OrchestrationCheckpointSummary struct {
	TurnId              TurnId                        `json:"turnId"`
	CheckpointTurnCount int                           `json:"checkpointTurnCount"`
	CheckpointRef       CheckpointRef                 `json:"checkpointRef"`
	Status              string                        `json:"status"`
	Files               []OrchestrationCheckpointFile `json:"files"`
	AssistantMessageId  *MessageId                    `json:"assistantMessageId,omitempty"`
	CompletedAt         string                        `json:"completedAt"`
}

// OrchestrationCheckpointFile describes a file changed in a checkpoint.
type OrchestrationCheckpointFile struct {
	Path      string `json:"path"`
	Kind      string `json:"kind"`
	Additions int    `json:"additions"`
	Deletions int    `json:"deletions"`
}

// OrchestrationThreadActivity is an activity event in a thread.
type OrchestrationThreadActivity struct {
	ID        EventId         `json:"id"`
	Tone      string          `json:"tone"`
	Kind      string          `json:"kind"`
	Summary   string          `json:"summary"`
	Payload   json.RawMessage `json:"payload"`
	TurnId    *TurnId         `json:"turnId,omitempty"`
	Sequence  *int            `json:"sequence,omitempty"`
	CreatedAt string          `json:"createdAt"`
}

File: mobile/composeApp/src/commonMain/kotlin/io/j3mobile/protocol/OrchestrationModels.kt
package io.j3mobile.protocol

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull

/** Top-level snapshot returned by getSnapshot. */
@Serializable
data class OrchestrationReadModel(
    val snapshotSequence: Int,
    val projects: List<OrchestrationProject>,
    val threads: List<OrchestrationThread>,
    val updatedAt: String,
)

/** A workspace project. */
@Serializable
data class OrchestrationProject(
    val id: ProjectId,
    val title: String,
    val workspaceRoot: String,
    val defaultModelSelection: ModelSelection? = null,
    val scripts: List<ProjectScript> = emptyList(),
    val createdAt: String,
    val updatedAt: String,
    val deletedAt: String? = null,
)

/** A user-defined script attached to a project. */
@Serializable
data class ProjectScript(
    val id: String,
    val name: String,
    val command: String,
    val icon: String,
    val runOnWorktreeCreate: Boolean,
)

/** A conversation thread. */
@Serializable
data class OrchestrationThread(
    val id: ThreadId,
    val projectId: ProjectId,
    val title: String,
    val modelSelection: ModelSelection,
    val runtimeMode: String,
    val interactionMode: String,
    val branch: String? = null,
    val worktreePath: String? = null,
    val latestTurn: OrchestrationLatestTurn? = null,
    val messages: List<OrchestrationMessage> = emptyList(),
    val proposedPlans: List<OrchestrationProposedPlan> = emptyList(),
    val activities: List<OrchestrationThreadActivity> = emptyList(),
    val checkpoints: List<OrchestrationCheckpointSummary> = emptyList(),
    val session: OrchestrationSession? = null,
    val createdAt: String,
    val updatedAt: String,
    val deletedAt: String? = null,
)

/** A single message in a thread. */
@Serializable
data class OrchestrationMessage(
    val id: MessageId,
    val role: String,
    val text: String,
    @Serializable(with = ChatAttachmentListSerializer::class)
    val attachments: List<ChatAttachment> = emptyList(),
    val turnId: TurnId? = null,
    val streaming: Boolean,
    val createdAt: String,
    val updatedAt: String,
)

object ChatAttachmentListSerializer : KSerializer<List<ChatAttachment>> {
    private val delegate = ListSerializer(ChatAttachment.serializer())

    override val descriptor: SerialDescriptor = delegate.descriptor

    override fun deserialize(decoder: Decoder): List<ChatAttachment> {
        val jsonDecoder = decoder as? JsonDecoder ?: return decoder.decodeSerializableValue(delegate)
        val element = jsonDecoder.decodeJsonElement()
        if (element is JsonNull) {
            return emptyList()
        }
        return jsonDecoder.json.decodeFromJsonElement(delegate, element)
    }

    override fun serialize(encoder: Encoder, value: List<ChatAttachment>) {
        encoder.encodeSerializableValue(delegate, value)
    }
}

/** A file attachment on a message. */
@Serializable
data class ChatAttachment(
    val type: String,
    val id: String,
    val name: String,
    val mimeType: String,
    val sizeBytes: Int,
)

/** Tracks the most recent turn in a thread. */
@Serializable
data class OrchestrationLatestTurn(
    val turnId: TurnId,
    val state: String,
    val requestedAt: String,
    val startedAt: String? = null,
    val completedAt: String? = null,
    val assistantMessageId: MessageId? = null,
)

/** Describes the runtime session for a thread. */
@Serializable
data class OrchestrationSession(
    val threadId: ThreadId,
    val status: String,
    val providerName: String? = null,
    val runtimeMode: String,
    val activeTurnId: TurnId? = null,
    val lastError: String? = null,
    val updatedAt: String,
)

/** A plan the assistant has proposed. */
@Serializable
data class OrchestrationProposedPlan(
    val id: String,
    val turnId: TurnId? = null,
    val planMarkdown: String,
    val implementedAt: String? = null,
    val implementationThreadId: ThreadId? = null,
    val createdAt: String,
    val updatedAt: String,
)

/** Describes a checkpoint in a thread. */
@Serializable
data class OrchestrationCheckpointSummary(
    val turnId: TurnId,
    val checkpointTurnCount: Int,
    val checkpointRef: CheckpointRef,
    val status: String,
    val files: List<OrchestrationCheckpointFile> = emptyList(),
    val assistantMessageId: MessageId? = null,
    val completedAt: String,
)

/** Describes a file changed in a checkpoint. */
@Serializable
data class OrchestrationCheckpointFile(
    val path: String,
    val kind: String,
    val additions: Int,
    val deletions: Int,
)

/** An activity event in a thread. */
@Serializable
data class OrchestrationThreadActivity(
    val id: EventId,
    val tone: String,
    val kind: String,
    val summary: String,
    val payload: JsonElement? = null,
    val turnId: TurnId? = null,
    val sequence: Int? = null,
    val createdAt: String,
)
